### ISSUE TITLE
How to specify Proxy Settings

### ISSUE THREAD

**Issue by epignosisx** [Microsoft/ApplicationInsights-aspnetcore#441]
We were trying out Application Insights at work and we are not getting any telemetry in the Azure portal other than Page Views. We are pretty sure it is related to our Corporate proxy (it has caused us high levels of pain in the past).

How can I specify the proxy settings? The default proxy will not cut it for us. We have to specify url, port, username and password for our proxy.

For our .NET Core we have to set it to the **HttpClientHandler** like this:

```cs
var httpClientHandler = new HttpClientHandler {
    Proxy = new MyProxy(), //our implementation of IWebProxy
};
```

I couldn't find the code in this repo in charge of sending the telemetry but if you are using **HttpClient** then it would be something similar.

Note that we are planning to use Application Insights in house, not in Azure, so we have the proxy restriction in all our environments.

**Comment by dnduffy** [comment]
Transmission is handled by the [base SDK](https://github.com/Microsoft/ApplicationInsights-dotnet).
Which specific class depends on whether you are using the InMemoryChannel or the ServerTelemetryChannel.

**Comment by epignosisx** [comment]
Thanks for pointing me in the right direction. I'm using the **ServerTelemetryChannel**. Looks like the issue is in this [line](https://github.com/Microsoft/ApplicationInsights-dotnet/blob/develop/src/Core/Managed/Shared/Channel/Transmission.cs#L62).

Ideally, I would like to be able to provide Application Insights with an ```HttpClient``` to use or at least be able to participate in the creation of the ```HttpClient``` so I can set the Proxy of the ```HttpClientHandler```.

Should I move this issue to the ApplicationInsights-dotnet repo?

**Comment by dnduffy** [comment]
@epignosisx Yes, please feel free to move it there, and if you want to submit a pull request that would be awesome!

**Comment by crippe** [comment]
@dnduffy | @epignosisx 

What is the status of this issue? Is there a new code that can handle proxy configurations? We are also governed by a proxy, though we use .NET 4.5 and then it seems like CreateRequest (https://github.com/Microsoft/ApplicationInsights-dotnet/blob/develop/src/Core/Managed/Shared/Channel/Transmission.cs#L374) is used today.

If no one had the opportunity to do something, can/should I make the changes and add a pull request?

In order to simplify, there should be something like:

```c#
if (this.UseProxy)
{
    request.Proxy = new WebProxy (this.ProxyAddress, this.ProxyBypassOnLocal);
}
```

**Comment by dnduffy** [comment]
@crippe Are you able to get the channel from dependency injection and modify it?  If not then please feel free to submit a pull request in that repo and we'll plan to take it in the next version.

**Comment by caretro** [comment]
Hi, any update on this issue? Is not it still possible to handle proxy configuration?
We are governed by a proxy too. We are using the ServerTelemetryChannel.

**Comment by crippe** [comment]
Shortly after my comment, it was decided centrally to use technology other than Application Insights. I'm sorry for the decision, but I could not do anything about it. Therefore, I did not make a request.

**Comment by caretro** [comment]
Hi, any update on this issue. Seems that the milestone is moved forward continuously...

**Comment by tremblaysimon** [comment]
Since HttpClient uses system proxy, I think it's possible to use a tool like cntlm to store the proxy credentials and configure a system proxy without a username/password.

**Comment by wernerbarnard** [comment]
Any update on this issue, or an alternative example to use AI with a proxy on-prem?

**Comment by cijothomas** [comment]
@wernerbarnard If you are configuring a system wide proxy, then SDK would just pick it up. If you'd like to configure proxy settings to applicationinsights channel - that is not supported today. I'll tag this as an enhancement.

**Comment by davebally** [comment]
I too am hitting this issue using AI on .Net Core behind  proxy.  Im not convinced that HttpClient in .NET Core does use the system proxy either so need to be able to set the proxy explicitly on AI to resolve.

**Comment by cijothomas** [comment]
@davebally 
Please run the following in a .NET Core Console app, and if it respects system proxy. If it does, then it'll print the AppId for the given ikey.

```
class Program
    {
        static void Main(string[] args)
        {
            HttpClient client = new HttpClient();
            var res = client.GetStringAsync("https://dc.services.visualstudio.com/api/profiles/your_instrumentationkey/appId").Result;
            Console.WriteLine(res);
        }       
    }
```

**Comment by davebally** [comment]
@cijothomas Thanks,  i retract that statement.  Looks like all is well,  i can see the call coming through in fiddler without an explicit proxy being set.

**Comment by nulltoken** [comment]
@cijothomas Unfortunately, this doesn't seem to work in all cases.

Running this code in our corporate environment raises a 407 error.

![image](https://user-images.githubusercontent.com/92363/58064918-3fee6b80-7b84-11e9-8c82-3b03cd47d426.png)

In our case, to make this pass, we have to define a `HttpHandler`, value the `Proxy` and `UseProxy` properties, and build the `HttpClient` from it. From a dev environment standpoint, that relies on running a local cntlm proxy, and configuring the `Proxy` property so that it points to it.

Teaching AppInsights to accept a `HttpHandler` would be very helpful for us.

**Comment by cijothomas** [comment]
@nulltoken Yes we are tracking this as an enhancement - to modify to accept httpclient settings.

**Comment by davebally** [comment]
@cijothomas Can you confirm that the HttpClient that is returned by default from IHttpClientFactory should be using the proxy ?  When i set the proxy with the below code i see the call come through in fiddler,  if i dont , i dont 

Thanks.

```
in startup ....

 services.AddHttpClient("external", c => { c.DefaultRequestHeaders.CacheControl = GetCacheControlHeader(); })
                .ConfigureHttpMessageHandlerBuilder(c => c.PrimaryHandler = GetHttpClientHandler(true));

  private static HttpClientHandler GetHttpClientHandler(bool isExternal = false)
        {
            var handler = new HttpClientHandler
            {
                AutomaticDecompression = System.Net.DecompressionMethods.GZip,
                SslProtocols = SslProtocols.Tls12,
                ClientCertificateOptions = ClientCertificateOption.Manual
            };

            if (isExternal)
            {

                var proxy = GetConfigProperty("Config", "Proxy", "Uri");
                if (!string.IsNullOrEmpty(proxy.Value))
                {
                    handler.Proxy = new WebProxy()
                    {
                        Address = new Uri(proxy.Value)
                    };
                }
            }
            else { 

                handler.ServerCertificateCustomValidationCallback = (message, cert, chain, errors) => true;
            }
            return handler;
        }

```

**Comment by cijothomas** [comment]
@davebally Sorry, i dont have any expertise in that area :( Its best to ask this in HttpClient github repo.

**Comment by wernerbarnard** [comment]
@davebelly - for me running inside docker with HTTP_PROXY and HTTPS_PROXY set as env variables, `new HttpClient` did NOT respect the proxy, but a HttpClient created from `IHttpClientFactory` did.

**Comment by cijothomas** [comment]
This is not planned for the next release 2.8.

**Comment by ArchonMegalon** [comment]
Is it planned at all? Having exactly the same problem here...

**Comment by cijothomas** [comment]
It is, and tagged with Future milestone. This means its not likely to land in next 2 stable releases. Please upvote the issues here so we promote items from Future to nearer milestones.

**Comment by mvdburght** [comment]
The following worked for us. Run it in the StartUp of your application:
``` csharp
private static void ConfigureProxyForApplicationInsightsHack(IWebProxy proxy)
{
    System.Runtime.CompilerServices.RuntimeHelpers.RunClassConstructor(typeof(Transmission).TypeHandle);
    var field = typeof(Transmission).GetField("client", BindingFlags.Static | BindingFlags.NonPublic);
    field.SetValue(null, new HttpClient(new HttpClientHandler { Proxy = proxy, UseProxy = true }));
}
```

**Comment by cijothomas** [comment]
@mvdburght thanks for sharing. The enhancement we are tracking here is to allow users to supply HttpClient for the channel/transmission.

**Comment by Swellenator** [comment]
I was getting a `407 Proxy Authorisation Required` type message. (which by the way I had to use WireShark to discover) I got it to work using @mvdburght 's workaround. My code is using a console app so I have this at the start of my `main` method:

```
 var proxy = Settings.GetSection("Proxy").Get<WebProxy>();

            if (!string.IsNullOrWhiteSpace(proxy.Address.ToString()))
            {
                var handler = new HttpClientHandler
                {
                    Proxy = proxy
                };

                System.Runtime.CompilerServices.RuntimeHelpers.RunClassConstructor(typeof(Transmission).TypeHandle);
                var field = typeof(Transmission).GetField("client", BindingFlags.Static | BindingFlags.NonPublic);
                field.SetValue(null, new HttpClient(handler));
            }
```

and in my config:

```
  "Proxy": {
    "Address": "http://[Url]:[Port]",
    "UseDefaultCredentials": true
  },
```

Hope that helps someone. Would be good to have an official way of doing it.
